### PR TITLE
Add `Nan` handling in the `GpuMin`

### LINKED ISSUE #6164: [FEA] Add `Nan` handling in the `GpuMin`
**Is your feature request related to a problem? Please describe.**
Currently, the `min` kernel will fall back to CPU if `hasNan = true`. We should let the `GpuMin` support the `hasNan` situation.
This is a sub task of #5320.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>
Close #6164.

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 1]
```diff
@@ -351,6 +351,11 @@ class CudfMin(override val dataType: DataType) extends CudfAggregate {
   override val name: String = "CudfMin"
 }
 
+/** Check if all values in a boolean column are trues */
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:354]: Same comment as for `CudfAny`. It would be good to have a comment explaining why we are using `CudfMin(BooleanType)`

**Review by revans2 — COMMENTED** [review]
Just needs tests and more comments to explain what is happening like the code for GpuMax.

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala (hunk 1 above)
Comment (HaoYang670) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:363]: Updated.

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 2]
```diff
@@ -624,18 +624,18 @@ def test_decimal128_min_max_group_by(data_gen):
...
+def test_float_min_max_reduction_all_nans(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
-        lambda spark: unary_op_df(spark, data_gen).selectExpr('max(a)'))
+        lambda spark: unary_op_df(spark, data_gen).selectExpr('min(a)', 'max(a)'))
 
 @ignore_order(local=True)
-@pytest.mark.parametrize('data_gen', [float_gen, double_gen], ids=idfn)
-def test_float_max_group_by_with_nan(data_gen):
+@pytest.mark.parametrize('data_gen', [float_gen, double_gen, SetValuesGen(FloatType(), [math.nan, None]), SetValuesGen(DoubleType(), [math.nan, None])], ids=idfn)
+def test_float_min_max_group_by_with_nan(data_gen):
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:634]: This is similar. Would it be better to just update the group by tests that use min and max to include float_gen and double gen.  Then we can also have some hard coded data to handle the different corner cases more like a unit test.

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 3]
```diff
@@ -624,18 +624,18 @@ def test_decimal128_min_max_group_by(data_gen):
             .agg(f.min('b'), f.max('b')))
 
 @ignore_order(local=True)
-@pytest.mark.parametrize('data_gen', [float_gen, double_gen], ids=idfn)
-def test_float_max_reduction_with_nan(data_gen):
+@pytest.mark.parametrize('data_gen', [SetValuesGen(t, [math.nan, None]) for t in [FloatType(), DoubleType()]], ids=idfn)
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:627]: What are we doing for structs that have NaNs in them?  The code shows that this is allowed, but I don't see tests for this.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 4]
```diff
@@ -522,6 +539,119 @@ case class GpuMin(child: Expression) extends GpuAggregateFunction
...
+      ),
+      updateMinVal.attr
+    )
+  )
+
+  // Same logic as the `inputProjection` stage.
+  override lazy val preMerge: Seq[Expression] = Seq (
+    GpuOr(GpuIsNan(evaluateExpression), GpuIsNull(evaluateExpression)),
+    GpuIsNan(evaluateExpression),
+    GpuNansToNulls(evaluateExpression)
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:617]: Can we have something to explain why GpuNansToNulls is needed here?

**Review by revans2 — DISMISSED** [review]
Looks good, just a few nits

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 2 above)
Comment (HaoYang670) [reply integration_tests/src/main/python/hash_aggregate_test.py]: updated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/nullExpressions.scala [hunk 5]
```diff
@@ -140,6 +140,20 @@ case class GpuIsNan(child: Expression) extends GpuUnaryExpression with Predicate
     input.getBase.isNan
 }
 
+case class GpuNansToNulls(child: Expression) extends GpuUnaryExpression{
```
[the remark this answers is not in the document]
Comment (HaoYang670) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/nullExpressions.scala:149]: updated

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala (hunk 4 above)
Comment (HaoYang670) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:621]: updated

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 3 above)
Comment (HaoYang670) [reply integration_tests/src/main/python/hash_aggregate_test.py]: > What are we doing for structs that have NaNs in them?

Sorry, which `structs`?

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 6]
```diff
@@ -624,18 +624,18 @@ def test_decimal128_min_max_group_by(data_gen):
...
+def test_float_min_max_reduction_all_nans(data_gen):
```
[the remark this answers is not in the document]
Comment (HaoYang670) [reply integration_tests/src/main/python/hash_aggregate_test.py]: updated.

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
Hi @revans2, could we merge this?

**Comment by HaoYang670** [comment]
Update the tests
```
test_min_max_for_single_level_struct
test_distinct_count_reductions
test_collect_list_reductions
```

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by HaoYang670** [comment]
build
